src/pages/posts/hooks/useObserver.ts

```diff
@@ -0,0 +1,30 @@
+import { useRef, useCallback, useState } from 'react';
+
+const useObserver = (callback: () => void) => {
+  const observer = useRef(
+    new IntersectionObserver(
+      (entries, _) => {
+        entries.forEach((entry) => {
+          if (entry.isIntersecting) {
+            if (observer !== null) {
+              observer.current.disconnect();
+            }
+            callback();
+          }
+        });
+      },
+      {
+        threshold: 1
+      }
+    )
```

Comment:
p1;
```suggestion
  const intersectionObserver = new IntersectionObserver(
    (entries, observer) => {
      entries.forEach((entry) => {
        if (entry.isIntersecting && observer) {
          observer.disconnect();
        } else if (entry.isIntersecting && !observer) {
          callback();
        } else {
          //else처리
        }
      });
    },
    {
      threshold: 1
    }
  );

  const observer = useRef(intersectionObserver);
```
이렇게 depth가 늘어나면 변수를 분리해주는게 더 가독성 좋을것 같아요!

Reply:
이렇게 분리하게 되는 경우 상위 컴포넌트가 리렌더링 될 때 마다 new IntersectionObserver 가 또 다시 생성된다는 문제가 있을 거 같아요...!

Reply:
```tsx
import { useRef, useCallback } from 'react';

const useObserver = (callback: () => void) => {
  const observerOption = useRef({ threshold: 1 });
  const observerFn: IntersectionObserverCallback = useCallback(
    (entries, observer) => {
      entries.forEach((entry) => {
        if (entry.isIntersecting && observer) {
          observer.disconnect();
        }
        if (entry.isIntersecting && !observer) {
          callback();
        }
      });
    },
    []
  );
  const observer = useRef(
    new IntersectionObserver(observerFn, observerOption.current)
  );

  const observe = useCallback((element: HTMLElement) => {
    if (element !== null) {
      observer.current.observe(element);
    }
  }, []);
  return [observe];
};

export default useObserver;

```

이렇게 옵션과 함수를 따로 빼내서 해결했습니다! 다음엔 가독성도 신경써야겠어요!